### PR TITLE
Graph theory - closed walks as word (12)

### PR DESCRIPTION
**Pull request by avekens**
This is the follow up PR of #2533, mainly replacing the local definition `*.f $e |- F = ( v e. V , n e. NN |-> { w e. ( n ClWWalksN G ) | ( w `` 0 ) = v } ) $.` (closed walks anchored at a fixed vertex and of a fixed length as word, see ~ numclwwlkovf, etc.) by the definition ClWWalksOnN. The second issue in #2533 (defining ClWWalksN, the set of closed walks of a fixed length as word, as function over NN0 instead of NN) will be solved in my next PR (also some TODO-AV's will be done then). 

Here the changes in detail:

main set.mm - misc.:
* theorem ~rabbii moved from PM's mathbox
* theorems ~annotanannot, ~dvdsmod0 added

main set.mm - section 5.7 "Words over a set"
* theorems ~ccat0, ~ccatws1clv, ~ccatws1lenp1b about concatenations of words added
* theorems ~ccatws1len, ~wrdlenccats1lenm1, ~ccatw2s1len, ~ccatws1n0
about the length of concatenations of words with singleton words revised: the alphabet of the singleton does not matter

main set.mm - Part 16 GRAPH THEORY
* header comment of section " Walks, paths and cycles" revised/enhanced
* theorems ~wwlknprcl, ~wwlknbp1, ~wwlknlsw added
* proofs of theorems ~wwlknbp2, ~clwlkclwwlk2, ~wwlksext2clwwlk shortened
* theorems ~clwwlknclwwlkdif and ~clwwlknclwwlkdifnum revised
* symbol ClWWalksOnN changed to ClWWalksNOn (analogous to WWalksNOn), label fragment "clwwlksonn" changed to "clwwlksnon" accordingly
* section "Closed walks as words" split into three subsections (as proposed by BJ) "Closed walks as words", "Closed walks of a fixed length as words" and "Closed walks on a vertex of a fixed length as words"; some theorems rearranged (moved into the appropriate subsection)
* theorems ~clwwlkson2x, ~clwwlkswwlksb, ~clwwlksnwwlksnb, ~clwwlksnonwwlknonb added
* theorems ~clwwlksndisj, ~clwwlksnun, ~clwwlksvbij revised
* theorems ~numclwwlkov* using local definition of F, corresponding to `( ClWWalksNon `` G )`, deleted and replaced by corresponding theorems using `( ClWWalksNon `` G )` in ~extwwlkfab, ~extwwlkfab2, ~extwwlkfabel, ~ numclwlk1lem2foa, ~numclwlk1lem2*, ~numclwwlk1,  ~numclwwlkqhash, ~numclwwlk2, ~numclwwlk3*, ~numclwwlk4, ~numclwwlk5
* proofs of theorems ~numclwwlk6, ~numclwwlk8  shortened

TA's mathbox:
* proof of ~sseqf shortened

AV's mathbox:
* comment of ~3an4ancom24 revised
* ~4an21 added

### CODE REVIEW

File: set.mm
Comment (tirix) [inline set.mm:40360]: This is only one step longer than `~rabiia`, is it sufficient to promote it?

File: set.mm
Comment (tirix) [inline set.mm:490363]: You seem to actually have made this proof longer - am I correct?

Comment (avekens) [reply set.mm:490363]: > You seem to actually have made this proof longer - am I correct?

Yes, you are right, because I made some ccat* theorems more general. Maybe the proof can become shorter if it is revised manually (I only changed the modified theorems, and ran minimize afterwards).

File: set.mm
Comment (avekens) [reply set.mm:40360]: I ran the minimize script for it, and it found 71 proofs which could be minimized (at least the number of bytes). Therefore, I think we can have it in main set.mm. I often looked for it an did not find it, so I was happy to find it in PM's mathbox now.

Comment (tirix) [reply set.mm:40360]: I see! Then of course it makes sense.

File: set.mm
Comment (tirix) [reply set.mm:490363]: Oh, I see: you changed the ccat* theorems in main, and that proof in my mathbox uses them, so it had to be adapted. Ok then, this is fine to me. 
...even though it is a bit strange that _removing antecedents_ makes that proof longer.

**Review by tirix — APPROVED** [review]
It's _a lot_ of changes, and so it's hard to review in details. 
But it is bound to be like that for this kind of refactoring.
I have not taken the time to review all details but I see what's being done here, with the change of local definition and the introduction of the new definition.

**Comment by avekens** [comment]
I think a line by line review is only useful up to line 419636, the rest is mainly restructuring or modifications of proofs.

File: set.mm
Comment (tirix) [inline set.mm:416923]: This theorem does not seem to bring in much compared to the previous one.

File: set.mm
Comment (tirix) [inline set.mm:183091]: This statement is stronger than `~dvdsval3`, I guess it may minimize theorems using `~dvdsval3`, did you check that?

**Comment by tirix** [comment]
Ok! I've added a pair of comments, but I still think this is good to merge. 
Shall we wait for a second approval or do you want to merge now?

File: set.mm
Comment (avekens) [reply set.mm:416923]: Yes, you are right, but ~wwlknbp2 is the older theorem which is used 11 times. I have to check each of these 11 theorems if their proofs can be shortened by using ~wwlknbp1, or at least do not become longer. If this is the csase, I will remove ~wwlknbp2. I put it on my TODO list.

File: set.mm
Comment (avekens) [reply set.mm:183091]: There are 25 proofs using ~dvdsval3, but none of them can be minimized (automatically). Maybe some of them can be shortend manually. I will pick some and have a look (-> my TODO- list)

**Comment by avekens** [comment]
Maybe @benjub wants to have a look at it? I would propose we wait until sunday evening, but I would like to have this PR merged on monday latest.

**Comment by benjub** [comment]
Sorry, I haven't had much time for reviews lately. I think a PR by @avekens reviewed by @tirix is two good reasons to be confident it meets set.mm's standards.

**Pull request merged**
